Ed: add two fusio files to import associated calendars

eturck opened the pull request:
Add two new files to fusio to explicitly link trips to certain grid calendars.
grid_rel_calendar_trip.txt works the same way as grid_rel_calendar_line.txt by
referencing a trip by its external code and associating it to the grid calendar referenced
by its id.
grid_rel_calendar_exception_dates.txt works like grid_exception_dates.txt and list
exceptions for each trip / grid_calendar association.
In build_associated_calendars, during data completion at the end of the import
process, when a meta_vj has been explicitly linked to one or more grid calendars via the fusio
files it will be ignored (a WARN log will mention how many has been ignored at the end).

Commit:
Ed: add two fusio files to import associated calendars

Add two new files to fusio to explicitly link trips to certain grid calendars.
grid_rel_calendar_trip.txt works the same way as grid_rel_calendar_line.txt by
referencing a trip by its external code and associating it to the grid calendar referenced
by its id.
grid_rel_calendar_exception_dates.txt works like grid_exception_dates.txt and list
exceptions for each trip / grid_calendar association.
In build_associated_calendars, during data completion at the end of the import
process, when a meta_vj has been explicitly linked to one or more grid calendars via the fusio
files it will be ignored (a WARN log will mention how many has been ignored at the end).

buildbotctp commented:
Can one of the admins verify this patch?

buildbotctp commented:
Can one of the admins verify this patch?

antoine-de commented:
Seems nice, thanks @eturck !

antoine-de commented:
Jenkins, test this please

antoine-de commented on source/ed/connectors/fusio_parser.cpp:
```diff
@@ -978,6 +982,83 @@ void CalendarLineFusioHandler::handle_line(Data&, const csv_row& row, bool is_fi
+    }
+    boost::gregorian::date date(parse_date(row[date_c]));
+    if(date.is_not_a_date()) {
+        LOG4CPLUS_ERROR(logger, "GridCalendarTripExceptionDatesFusioHandler: Date format not valid, we do not add the exception " <<
```
since it does not stop the reading, I think the WARN level is enough

antoine-de commented on source/ed/connectors/fusio_parser.cpp:
```diff
@@ -978,6 +982,83 @@ void CalendarLineFusioHandler::handle_line(Data&, const csv_row& row, bool is_fi
+    }
+    auto meta_vj = gtfs_data.metavj_by_external_code.find(row[trip_c]);
+    if (meta_vj == gtfs_data.metavj_by_external_code.end()) {
+        LOG4CPLUS_ERROR(logger, "CalendarTripFusioHandler: Impossible to find the trip " << row[trip_c]);
```
same here for the log level

antoine-de commented on source/ed/connectors/fusio_parser.cpp:
```diff
@@ -342,6 +342,10 @@ std::vector<ed::types::VehicleJourney*> TripsFusioHandler::get_split_vj(Data& da
     types::MetaVehicleJourney& meta_vj = data.meta_vj_map[row[trip_c]]; //we get a ref on a newly created meta vj
+    // Store the meta vj with its external code, only if this code exists
+    if(is_valid(ext_code_c, row)){
+        gtfs_data.metavj_by_external_code.insert({row[ext_code_c], &meta_vj});
```
why do you need this external code map ?

why don't you link the vj by it's id in the files ?

Commit:
Ed: change log level to warn during parsing of grid_calendar_trip_exception_dates in fusio_parser

eturck commented:
Thanks for the feedback!
I have changed the log level to WARN in GridCalendarTripExceptionDatesFusioHandler::handle_line, since it's indeed enough.

For the last part, I link the vehicle journey by its external code in the two files in order to keep consistency with grid_rel_calendar_line.txt. Since it's using "line_external_code" I thought it would make the most sense to use "trip_external_code".
But I'm open to change it if you don't want to add this external code map or if it's better to link it by its id.

antoine-de commented:
Jenkins, test this please

antoine-de commented on source/ed/data.cpp:
```diff
@@ -755,6 +762,9 @@ void Data::build_associated_calendar() {
         LOG4CPLUS_WARN(log, "no calendar found for " << nb_not_matched_vj << " vehicle journey");
     }
+    if (nb_ignored) {
+        LOG4CPLUS_WARN(log, nb_ignored << " vehicle journey were already linked and therefore ignored" );
```
I think INFO level is enough here

antoine-de commented:
ok, it seems it's easier on our side too to link with the external code instead of the vj's id (even if that does not seems very consistent with the rest of the format).

Apart from the last log level to change it seems perfect for me, thanks!

Commit:
Ed: Change log level to INFO in build_associated_calendars

Use INFO to report the number of vehicle journey ignored in build
associated calendars (vehicle journey explicitly linked to grid
calendars via grid_rel_calendar_trip.txt file)

eturck commented:
Changed it.

In the last NTFS specification you send to us I see that grid_rel_calendar_line can use line_external_code or line_id (in fusio_parser.cpp only line_external_code is used right now).
Do you want me to parse trip_id in addition to trip_external_code in CalendarTripFusioHandler::handle_line ? And maybe add the parsing of line_id for grid_rel_calendar_line too ?

antoine-de commented:
no I think it's better to link only on one ID, so it's ok for me for the link with external codes.
